plugins/out_datadog/datadog.c

```diff
@@ -0,0 +1,279 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019      The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_output.h>
+#include <fluent-bit/flb_io.h>
+#include <fluent-bit/flb_log.h>
+#include <fluent-bit/flb_http_client.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_time.h>
+
+#include <msgpack.h>
+
+#include "datadog.h"
+#include "datadog_conf.h"
+
+static int cb_datadog_init(struct flb_output_instance *ins,
+                           struct flb_config *config, void *data)
+{
+    struct flb_out_datadog *ctx = NULL;
+    (void) data;
+
+    ctx = flb_datadog_conf_create(ins, config);
+    if (!ctx) {
+        return -1;
+    }
+
+    /* Set the plugin context */
+    flb_output_set_context(ins, ctx);
+    return 0;
+}
+
+static int64_t timestamp_format(const struct flb_time* tms) {
+    int64_t timestamp = 0;
+    /* Format the time, use milliseconds precision not nanoseconds */
+    timestamp = tms->tm.tv_sec * 1000;
+    timestamp += tms->tm.tv_nsec/1000000;
+    /* round up if necessary */
+    if (tms->tm.tv_nsec % 1000000 >= 500000) {
+        ++timestamp;
+    }
+    return timestamp;
+}
+
+static void dd_msgpack_pack_key_value_str(msgpack_packer* mp_pck,
+                                          const char *key, size_t key_size,
+                                          const char *val, size_t val_size)
+{
+    msgpack_pack_str(mp_pck, key_size);
+    msgpack_pack_str_body(mp_pck, key, key_size);
+    msgpack_pack_str(mp_pck, val_size);
+    msgpack_pack_str_body(mp_pck,val, val_size);
+}
+
+static int datadog_format(const void *data, size_t bytes,
+                          const char *tag, int tag_len,
+                          char **out_data, size_t *out_size,
+                          struct flb_out_datadog *ctx)
+{
+    /* for msgpack global structs */
+    size_t off = 0;   
+    int array_size = 0;
+    msgpack_unpacked result;
+    msgpack_sbuffer mp_sbuf;
+    msgpack_packer mp_pck;
+    /* for sub msgpack objs */
+    int map_size;
+    struct flb_time tms;
+    int64_t timestamp;
+    msgpack_object *obj;
+    msgpack_object map;
+    msgpack_object root;
+    msgpack_object k;
+    msgpack_object v;
+    /* output buffer */
+    flb_sds_t out_buf;
+
+    /* Count number of records */
```

Comment:
to count the number of items, please use the new flb_mp_count() function, defined here:

https://github.com/fluent/fluent-bit/blob/master/src/flb_mp.c#L24